### PR TITLE
Rename OfflineVirtualMachine to StatefulVirtualMachine

### PR DESCRIPTION
**Pull request by mlsorensen**
This addresses #985 and is just a starting off point that passes my initial testing. I'm sure there are other pieces I'm not aware of that will need to be adjusted as well. 

* I'm not sure what will need to be changed in CI/testing. 
* I see at the very least we need to adjust kubevirt/demo.
* What is desired as far as extra documentation and/or release notes?

### CODE REVIEW

**Comment by stu-gott** [comment]
ci test please

**Comment by stu-gott** [comment]
Thanks @mlsorensen! nice job!

**Comment by rmohr** [comment]
ok to test

**Comment by mlsorensen** [comment]
The RBAC test should be fixed now.

Not sure what is up with the VM lifecycle test, looks like it's failing to find a virt-handler pod on one of the cluster nodes. I'm not certain if it's related to this change.

**Comment by cynepco3hahue** [comment]
@mlsorensen I think it does not relate, but all CI nodes must run `virt-handler` pod, and if for some reason it does not run, we need to investigate it and not just choose another node.

**Comment by mlsorensen** [comment]
@cynepco3hahue   If all CI nodes run virt-handler, then behavior should remain the same, it should choose the same host it always did. However, the integration tests as they are won't run on any old cluster, as many clusters are set up with 'master' nodes that don't have virt-handler, and these tests are hard-coded to choose the first node out of all nodes, regardless. This seemed wrong to me, as I think others would want to use these tests to validate KubeVirt on their clusters. I wasn't able to validate my RBAC fix in my environment without this patch, though perhaps I could have by spinning up a dev cluster.

I agree this may have the effect of masking a misconfiguration in CI cluster if the intention was that the CI cluster should have all schedulable nodes, but it doesn't seem that this test's purpose is to ensure the CI cluster config. I'm happy to revert this commit to get this change applied.

**Comment by mlsorensen** [comment]
Looks like I've got another merge conflict with the latest pull. Let me fix that. 

Please let me know what else needs to be done for this change, or if we want to hold off, as this is probably going to continue to cause conflicts as more pull requests come through.

**Comment by fabiand** [comment]
A little late to this party.

We had a few chats with people looking into KubeVirt at KubeCon and RH Summit.
One thing which comes up is that OVMs might be the things current users care about most.
Thus maybe a different pattern is also something to think about befor doing the change: Make OVMs the VM, and find a different name for VMs. Aka `VirtualMachine` and `VirtualMachineInstance` or alike.
I'll probably have some statistics at hand in a bit of time which is supporting this.

@mlsorensen what entity do you actually work with - VM or OVM?

**Comment by rmohr** [comment]
> One thing which comes up is that OVMs might be the things current users care about most.

That is very likely. If you don't run external controllers, you also almost never use a pod directly (except of mirrored pods).

> Thus maybe a different pattern is also something to think about befor doing the change: Make OVMs the VM, and find a different name for VMs. Aka VirtualMachine and VirtualMachineInstance or alike.

While i don't want to position myself right now, keep this in mind: `Instance` is always implied. It is a Pod and not a `PodInstance`. It is a `StatefulSet` and not a `StatefulSetInstance`. Further the StatefulVirtualMachine does something for you already, which also fits nice into the  k8s world, whereas VirtualMachine alone does not mean much, except that this is your live thing (again the equivalent of a "PodInstance").

> I'll probably have some statistics at hand in a bit of time which is supporting this.

Please also  include the questions and differentiate between the case where they demanded a name change or just meant that the StatfulVirtualMachine will be the one which they will use.

**Comment by mlsorensen** [comment]
@fabiand I believe most of our users will be interested in the stateful version. I have reservations about reusing VirtualMachine at this point, simply because of severe (yet temporary) confusion for users *and* within the codebase (seems tricky to go in and mass-fix everything there).  

I like the take @rmohr gives.

I guess there is some issue with storage in the test environments? vagrant-dev looks like it passed but the others seemed to have IO timeouts and/or complain about volume issues.

**Comment by davidvossel** [comment]
> We had a few chats with people looking into KubeVirt at KubeCon and RH Summit.
One thing which comes up is that OVMs might be the things current users care about most.
Thus maybe a different pattern is also something to think about befor doing the change: Make OVMs the VM, and find a different name for VMs. Aka VirtualMachine and VirtualMachineInstance or alike.
I'll probably have some statistics at hand in a bit of time which is supporting this.

VirtualMachines and StatefulVirtualMachines please. Our naming conventions here are following kubernete's lead, not other VM management systems. In k8s, ephemeral is assumed and stateful is something that is explicitly requested. 

I really don't want this to become more convoluted conversation. That's how we ended up with OfflineVirtualMachines to begin with.

**Comment by rmohr** [comment]
> VirtualMachines and StatefulVirtualMachines please. Our naming conventions here are following kubernete's lead, not other VM management systems. In k8s, ephemeral is assumed and stateful is something that is explicitly requested.

I see it the same way.

> I really don't want this to become more convoluted conversation. That's how we ended up with OfflineVirtualMachines to begin with.

+1 on proceeding with StatefulVirtualMachine.

**Comment by stu-gott** [comment]
I agree with @mlsorensen's point that at this point VirtualMachine is pretty strongly fossilized in people's minds. Thus I tend to also agree with @davidvossel and @rmohr that the correct course of action is to rename OfflineVirtualMachine to StatefulVirtualMachine.

**Comment by davidvossel** [comment]
@mlsorensen can you rebase/clean up this branch so only the stateful vm changes are visible in the diff? I'm seeing patches related to other PRs in here.

**Comment by mlsorensen** [comment]
Yeah, somehow I got those duplicate commits pushed when I was updating. I'm just going to push a fresh branch and abandon this, probably the easiest way to clean things up. See #1020.

**Comment by itamarh** [comment]
as to "everyone knows a VM as it is today" - I have a problem comparing our current set of contributors and users with the entire virtualization world that for years "fossilized" on a VM being, well, a VM, not the running instance which a pod reflects in Kubernetes. Hence we should consider this before we grow much more in adoption, and potentially confuse way more people.
We heard some comments about this last week. I wonder if there is a better way to "ask" about this from say the ovirt and OSP community (or another way).

**Comment by mlsorensen** [comment]
@itamarh It's a good point. I think the use of "VM" over the years has been a bit flexible anyway, for example EC2 VMs have had various degrees of "ephemeralness" over the years depending on how one deploys.

My comments about VirtualMachine being entrenched were more around the amount of effort to audit the code and comments to correctly replace every instance of 'vm', 'virtual machine', 'VirtualMachine', etc with something else, or avoiding the change, as appropriate. Also carrying that forward and keeping a clear distinction in the code and comment use of 'VM' to refer to the running instance vs 'VM' the stateful resource. Right now having 'VM' as both the base unit of deployment (the running 'thing') and the name of the resource that represents it makes things pretty straightforward.

There are good arguments in both directions. In the end there's only a minor learning curve to names, however we choose to use them. This branch was just an attempt to make the naming better reflect what the resource is, as either way 'OfflineVirtualMachine' always seemed to raise questions about what 'Offline' meant.

**Comment by davidvossel** [comment]
> We heard some comments about this last week. I wonder if there is a better way to "ask" about this from say the ovirt and OSP community (or another way).

@itamarh  cnv razor 

We're looking to k8s to define the structure and nomenclature for our abstractions not ovirt/osp.  This is the element of least surprise for everyone, both devs and users.

OVM is actually a perfect example of us departing from the CNV razor and the confusion that is generated as a result. StatefulVirtualMachine is us correcting that mistake.

**Comment by itamarh** [comment]
@mlsorensen I am trying to avoid confusion for "the rest of the world". I appreciate its more work for us. We could avoid it if the running entity is named clearly different (I'd say Instance, but its overloaded with VM in clouds already).
Even in clouds, originally meant to be ephemeral the Instance notation actually grew out to be exactly like a VM - it stays when you stop it, unless you delete/terminate it.
So basically everyone using VMs expects the entity they work with to stay when stopped.
What I would like to avoid is having to educate "the rest of the world" why we are philosophically correct about the naming we chose in KubeVirt to match Kubernetes (I can also argue the fact we got used to VM being the running entity in KubeVirt doesn't make it the correct naming).
When we look at a broader ecosystem to hopefully embrace and integrate with KubeVirt, all these products already work with existing virt/cloud solution, and most likely call them VMs. Now we'd need to tell everyone that the VM entity they see in the UI/API when they work with KubeVirt means something else...
So I really wish we had a better name for the "running VM" so everyone could perform actions like start/stop on the VM entity, like they expect to from (decades now) of virtualization experience.

**Comment by itamarh** [comment]
@davidvossel we are talking about nomenclature. it goes to what people expect. Kubernetes taught everyone pods are stateless, hence they needed StatefulSets.
Everyone expect VMs to be statefull already, hence us deciding to not use that name for something stateful is the confusing part to me.

**Comment by mlsorensen** [comment]
Understood. I do think there's some initial value for traditional users in having the basic 'VM' be some persistent definition. On the other hand, as I point out, VMs aren't always persistent even on today's popular management platforms. Granted, their configuration is, usually.

I don't think it's a problem so much for Kubernetes users, who already understand the concepts and went through such a shift with ideas like needing to wrap a Pod in a Deployment. I think it may be a minimal learning curve for non-Kubernetes natives, but might teach them something basic and important about Kubernetes in the process; ho higher order resources encapsulate and manage the basic units of work.

Is KubeVirt's goal more to compete with and replace traditional VM orchestration platforms, or is it to bring Kubernetes concepts and strengths to VM management? I totally understand the desire for customers who use traditional orchestration platforms to 'just want their VM', how much does this project want to hide Kubernetes ideas and just give them a VM?

**Comment by davidvossel** [comment]
> Everyone expect VMs to be statefull already, hence us deciding to not use that name for something stateful is the confusing part to me.

yes, I agree kubernetes is confusing when coming from traditional VM management platforms. The assumptions are reversed. I actually struggled with this coming from management of VMs in GCE and AWS to containers in Kubernetes.

Anyone who is coming from a traditional VM management platform to kubernetes/containers is having to deal with that learning curve regardless if they use KubeVirt or not. I believe our abstractions need to be consistently named within the ecosystem they exist in. For example, a StatefulVM might be strange in the context of the oVirt ecosystem, but in the context of Kubernetes it has a well understood meaning.

**Comment by itamarh** [comment]
@mlsorensen I think VMs are persistent even on clouds (named instances). i.e., they behave exactly like VMs behave (they started thinking differently, but ended in the same place).
The instance/vm is the persistent config. Which image to boot from is kept if defined. The disk/volume may or may not be persistent, but is kept. The elastic IP is associated with it, even if the disk/volume is not persistent, etc.
As to target audience, well, I'd love for anyone to find value in this of course. I would like to see the ecosystem around virt embracing and integrating with it. While we bring new concepts from Kubernetes to virt, even the cloud attempts to make VMs behave differently didn't work, and they became, mostly, the same stateful VMs. I am not sure in Kubernetes suddenly VMs would behave differently. If someone would invest in changing them, they probably would containerize them to begin with. But the ROI is not there. Which means, for many, the VMs are the same VMs, now used and using Kubernetes mechanisms. Now the backup vendors and VM management/provisioning solutions of the world integrate with it. It is just another virt solution to them and their admins. They would not change in the UI/API from VM. we would need to keep explaining to that entire ecosystem why we chose to be different. Every SDK out there abstracting VMs would need to explain to their users in KubeVirt the VM entity doesn't map to the VM entity in kubevirt, etc.
So the audience in my view is also the entire virt audience, a lot of which indeed may not care about kuebrnetes, as their role would be for the virt part in it.
basically, I'd expect almost any SDK / UI out there which will show VMs to users, to call the SVM a VM... I'm  sure even if we designed a dedicated UI for KubeVirt, we would call it a VM...
(of course we did: https://www.youtube.com/watch?v=J07dW5VZJtg )
This isn't changing anything in the actual implementation, we are just talking about naming here :)

**Comment by rmohr** [comment]
> So the audience in my view is also the entire virt audience, a lot of which indeed may not care about kuebrnetes, as their role would be for the virt part in it.

In this case they are probably then best of with very high-level abstractions like ManageIQ. As soon as they get closer to the CLI level, it will become visible that the StatefulVirtualMachine itself is not the "traditional" VM either. And while we can for sure cover every use-case one can come up with, I am not sure if we can or want mimic the classical, usual transaction-based system for them on that level.

> This isn't changing anything in the actual implementation, we are just talking about naming here :)

Well yes "just" ;). So even if we call our StatefulVirtualMachine a "VirtualMachine", it would still not be the same like in e.g. oVirt. In such systems the whole API is tailored to the thinking of the one object you interact with. In KubeVirt VirtualMachine provides a part of the usual functionality and StatefulVirtualMachine provides another part of that functionality. That separation and the functionalities added to each of these entities make a lot of sence when integrating with k8s (at least from my view). Wouldn't it be strange if we then call the other part of the functionality, which is even less the VirtualMachine finally the VirtualMachine? That also caused issues there. A lot of actions (when it e.g comes to hotplug vs. config changes on restart) are hard to model and interact with. That is one example where we can simply improve the situation. No matter if a user cares about k8s per-se or not.

I also want to add, that it is possible for a very high-level system to combine all these entities to its own personal view of what a VM means for it (thinking of ManageIQ here for instance). In that case the user/admin would either way never see a StatefulVirtualMachine, VirtualMachine or VirtualMachineInstence. It would just see its usual entity condensed out of all of that.

Just discussed that with @itamarh  on IRC, so not new for you that I think that, but interestingly such a condensed view on ManageIQ would be, from the KubeVirt perspective, more low-level than the KubeVirt entities themselves, since it would hide the "caring about the service which you provide with your vm"-thinking again.

**Comment by fabiand** [comment]
Just getting back from the weekend and there seems to be a bit to catch up with :)

The tension seems to arise from the fact that `VirtualMachine` is a good term to describe the "smallest" entity, but at the same time  a `VirtualMachine` is also a good term to describe a common virtual machine use-case.

My reasoning was that using `VirtualMachine` as a name for the current OVMs might make sense, because most people might want to use those (and not the current VMs).
However, such a rename would prevent us from using the `VirtualMachine` term for the smallest unit, and this seems to be the tension point - to then not to use the term VM for the smallest unit anymore.

(Btw - It was not my intention to introduce a VM_Instance_ it was merely a try to find a good name for the smallest unit).

I do see good reasons for both cases

- to use the term `VirtualMachine` for the common use-case (stateful VM like)
- to use the term `VirtualMachine` for the smallest unit

The options I am seeing:

a. VM for common use-case - find a better term for smallest unit (domain, some-smart-name)
b. VM for smallest unit - proceed with this PR, StatefulVM for common use-case
c. Make StatefulVM the smallest unit (hooray!)

Wrt c - We don't want to go with c for several reasons (i.e. OVM provides more than a VM).

Wrt b - I think it should be fine to go with a different term than V for the smallest unit i. because pods are also not containers ii. higher level kube workloads also do not contain the name of the smallest unit.

Wrt a - I think it's good to find the right solution for the users.

Personally I actually do like the structure of StatefulVM and VM. But still ponder on this topic because we probably all don't want to change it again in a few months.

Thoughts?

**Comment by davidvossel** [comment]
> Thoughts?

The goal here is to bring virtualization into the container ecosystem.

The tension in this conversation appears to be centered around us trying to bring user assumptions from ovirt/osp into the container space. That's the exact problem that lead to us picking OVM in the first place, so I want to encourage us to pick a naming convention that works within the kubernetes ecosystem. StatefulVirtualMachine does that. It's possible there is another option, but no one has suggested anything better yet.
  
If someone can suggest something better than VM for the smallest unit, then we can have the conversation of whether to make the OVM -> VM switch and VM -> (???) switch. I don't have a better solution than StatefulVirtualMachine and VirtualMachine. In the container ecosystem this naming convention accurately reflects our intent.

**Comment by mlsorensen** [comment]
+1, let’s stick to Kubernetes ideas. I don’t think it will be confusing to users to have StatefulVirtualMachine, it’s just a little different than some may be used to, and that’s OK.

**Comment by michalskrivanek** [comment]
I can see the argument that people not caring about UI will benefit from more aligned terminology and have no problems with this change, and UI people will not care about the names anyway as they would be abstracted. So I guess it's fine to rename. 
There's just one more thing to consider - we do have existing code depending on OVM - Openshift Console depends on it, and the current code is going to be released in 3.10. I expect we need to move fast if we want to make it (and someone has to do it too:)

**Comment by michalskrivanek** [comment]
> @itamarh I think it would be up to the abstraction layer above to present what it wants, which is fine. If it wants to use Kubernetes under the hood to fulfill VMs, it can name its user-facing entities whatever it wants. Going forward if we do VirtualMachineReplicaSets or StatefulVirtualMachineSets, a layer on top might just call this a "VM cluster" or something.

that's true to some degree, UIs like openshift-web-console try to reflect the underlying names. We already departed from that with "Virtual Machines" representing both OVMs and VMs since it just doesn't make logical sense to separate them in a view which is supposed to "list all your VMs". It's not very intuitive though...the filtering by entity type behaves differently than one might expect already.

> As a developer on a cloud VM orchestration platform, I don't agree that all products will treat the StatefulVirtualMachine as the de-facto VM, they might not even label them VMs. It may be 'compute' or something.
... which is the "de-facto VM" meaning. I believe all those products will basically ignore VM (whatever it's called) and deal with OVM(or whatever it's called) instead, calling them Compute perhaps. Or with replica sets calling them "Pools".

**Comment by fabiand** [comment]
Thanks for the work, closing this in favor of #1102
